File: BESmashClub/APIWebApp/Controllers/AuthController.cs
using Entites.DTOs.Auth;
using Entites.DTOs.Common;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Services.Implementations;
using Services.Interfaces;
using System.Security.Claims;

namespace APIWebApp.Controllers;

[ApiController]
[Route("api/auth")]
public class AuthController : ControllerBase
{
    private readonly IAuthService _authService;
    private readonly IEmailService _emailService;
    public AuthController(IAuthService authService, IEmailService emailService)
    {
        _authService = authService;
        _emailService = emailService;
    }

    private void SetRefreshTokenCookie(string refreshToken)
    {
        var cookieOptions = new CookieOptions
        {
            HttpOnly = true,
            Secure = true, 
            SameSite = SameSiteMode.None,
            Expires = DateTime.UtcNow.AddDays(7)
        };
        Response.Cookies.Append("refreshToken", refreshToken, cookieOptions);
    }

    /// <summary>
    /// Đăng ký tài khoản mới (cần xác thực email để kích hoạt).
    /// </summary>
    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        try
        {
            await _authService.RegisterAsync(request);
            return Ok(ApiResponse.SuccessResponse("Đăng ký thành công. Vui lòng xác thực email để kích hoạt tài khoản."));
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
    }

    /// <summary>
    /// Xác thực đăng ký qua mã OTP (kích hoạt tài khoản và tự động đăng nhập).
    /// </summary>
    [HttpPost("verify-registration")]
    public async Task<IActionResult> VerifyRegistration([FromBody] VerifyEmailRequest request)
    {
        try
        {
            var ipAddress = HttpContext.Connection.RemoteIpAddress?.ToString();
            var userAgent = HttpContext.Request.Headers.UserAgent.ToString();

            var result = await _authService.VerifyRegistrationAsync(request, ipAddress, userAgent);
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<TokenResponse>.SuccessResponse(result, "Xác thực và kích hoạt tài khoản thành công."));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (Exception ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
    }

    /// <summary>
    /// Gửi lại mã OTP xác nhận tài khoản.
    /// </summary>
    [HttpPost("resend-verification-code")]
    public async Task<IActionResult> ResendVerificationCode([FromBody] string email)
    {
        try
        {
            await _authService.ResendVerificationCodeAsync(email);
            return Ok(ApiResponse.SuccessResponse("Đã gửi lại mã xác nhận qua email. Vui lòng kiểm tra hộp thư."));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (Exception ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
    }

    /// <summary>
    /// Đăng nhập (trả về Access Token & Refresh Token).
    /// </summary>
    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        try
        {
            var ipAddress = HttpContext.Connection.RemoteIpAddress?.ToString();
            var userAgent = HttpContext.Request.Headers.UserAgent.ToString();

            var result = await _authService.LoginAsync(request, ipAddress, userAgent);
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<TokenResponse>.SuccessResponse(result, "Đăng nhập thành công."));
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(ApiResponse.ErrorResponse(ex.Message));
        }
    }

    /// <summary>
    /// Cấp lại Access Token mới dựa trên Refresh Token.
    /// </summary>
    [HttpPost("refresh-token")]
    public async Task<IActionResult> RefreshToken()
    {
        try
        {
            var refreshToken = Request.Cookies["refreshToken"];
            if (string.IsNullOrEmpty(refreshToken))
                return Unauthorized(ApiResponse.ErrorResponse("Không tìm thấy Refresh Token trong cookie."));

            var ipAddress = HttpContext.Connection.RemoteIpAddress?.ToString();
            var userAgent = HttpContext.Request.Headers.UserAgent.ToString();

            var result = await _authService.RefreshTokenAsync(refreshToken, ipAddress, userAgent);
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<TokenResponse>.SuccessResponse(result, "Token đã được làm mới."));
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(ApiResponse.ErrorResponse(ex.Message));
        }
    }

    /// <summary>
    /// Đăng xuất (vô hiệu hóa Refresh Token hiện tại).
    /// </summary>
    [Authorize]
    [HttpPost("logout")]
    public async Task<IActionResult> Logout()
    {
        var refreshToken = Request.Cookies["refreshToken"];
        if (string.IsNullOrEmpty(refreshToken))
            return BadRequest(ApiResponse.ErrorResponse("Không tìm thấy Refresh Token."));

        var userId = Guid.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        await _authService.LogoutAsync(userId, refreshToken);
        Response.Cookies.Delete("refreshToken", new CookieOptions { HttpOnly = true, Secure = true, SameSite = SameSiteMode.None });
        return Ok(ApiResponse.SuccessResponse("Đăng xuất thành công."));
    }

    // ===================== PASSWORD RESET =====================

    /// <summary>
    /// Gửi email đặt lại mật khẩu.
    /// </summary>
    [HttpPost("forgot-password")]
    public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordRequest request)
    {
        try
        {
            await _emailService.SendPasswordResetAsync(request.Email);
            return Ok(ApiResponse.SuccessResponse("Đã gửi email đặt lại mật khẩu. Vui lòng kiểm tra hộp thư."));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (Exception ex)
        {
            return StatusCode(500, ApiResponse.ErrorResponse("Không thể gửi email. Vui lòng thử lại sau."));
        }
    }

    /// <summary>
    /// Đặt lại mật khẩu bằng mã code từ email.
    /// </summary>
    [HttpPost("reset-password")]
    public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordRequest request)
    {
        try
        {
            await _emailService.ResetPasswordAsync(request.Code, request.Email, request.NewPassword);
            return Ok(ApiResponse.SuccessResponse("Đặt lại mật khẩu thành công."));
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ApiResponse.ErrorResponse(ex.Message));
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ApiResponse.ErrorResponse(ex.Message));
        }
    }
}

File: AndroidApp/lib/src/features/auth/presentation/controllers/auth_controller.dart
import 'package:flutter/material.dart';
import '../../domain/repositories/auth_repository.dart';
import '../../data/models/auth_models.dart';
import '../../../../shared/network/api_response.dart';
import '../../../../shared/widgets/global_loading_overlay.dart';

/// Bộ điều khiển trạng thái xác thực (Auth Controller) ở lớp Presentation.
/// Chịu trách nhiệm gọi tầng Nghiệp vụ (Repository) và quản lý hiệu ứng hiển thị Loading Overlay.
class AuthController {
  final AuthRepository authRepository;

  AuthController({required this.authRepository});

  /// Thực hiện tác vụ Đăng nhập hệ thống.
  /// Nhận thông tin đăng nhập, gọi API từ Repo và tự động xử lý vòng đời của Loading Overlay.
  Future<ApiResponse<TokenResponse>> login(
    BuildContext context, {
    required String email,
    required String password,
  }) async {
    // Hiển thị loading overlay chặn tương tác người dùng
    GlobalLoading.show(context, message: 'Đang đăng nhập...');

    try {
      final request = LoginRequest(email: email, password: password);
      final response = await authRepository.login(request);
      return response;
    } finally {
      // Luôn đóng loading overlay bất kể thành công hay thất bại
      GlobalLoading.dismiss();
    }
  }

  /// Thực hiện tác vụ Đăng ký tài khoản mới.
  /// Nhận thông tin đăng ký, gọi API từ Repo và tự động xử lý vòng đời của Loading Overlay.
  Future<ApiResponse<void>> register(
    BuildContext context, {
    required String fullName,
    required String email,
    required String password,
    required String phoneNumber,
  }) async {
    // Hiển thị loading overlay chặn tương tác người dùng
    GlobalLoading.show(context, message: 'Đang tạo tài khoản...');

    try {
      final request = RegisterRequest(
        fullName: fullName,
        email: email,
        password: password,
        phoneNumber: phoneNumber,
      );
      final response = await authRepository.register(request);
      return response;
    } finally {
      // Luôn đóng loading overlay bất kể thành công hay thất bại
      GlobalLoading.dismiss();
    }
  }

  /// Xác thực OTP khi vừa đăng ký xong.
  Future<ApiResponse<TokenResponse>> verifyRegistration(
    BuildContext context, {
    required String email,
    required String code,
  }) async {
    GlobalLoading.show(context, message: 'Đang xác thực...');
    try {
      final request = VerifyEmailRequest(email: email, code: code);
      return await authRepository.verifyRegistration(request);
    } finally {
      GlobalLoading.dismiss();
    }
  }

  /// Gửi lại mã OTP qua email (trong màn hình đăng ký).
  Future<ApiResponse<void>> resendVerificationCode(
    BuildContext context, {
    required String email,
  }) async {
    GlobalLoading.show(context, message: 'Đang gửi lại mã...');
    try {
      return await authRepository.resendVerificationCode(email);
    } finally {
      GlobalLoading.dismiss();
    }
  }
  /// Gửi yêu cầu lấy lại mật khẩu qua email.
  Future<ApiResponse<void>> forgotPassword(
    BuildContext context, {
    required String email,
  }) async {
    GlobalLoading.show(context, message: 'Đang gửi yêu cầu...');
    try {
      return await authRepository.forgotPassword(email);
    } finally {
      GlobalLoading.dismiss();
    }
  }

  /// Đặt lại mật khẩu mới.
  Future<ApiResponse<void>> resetPassword(
    BuildContext context, {
    required String email,
    required String code,
    required String newPassword,
  }) async {
    GlobalLoading.show(context, message: 'Đang khôi phục...');
    try {
      final request = ResetPasswordRequest(
        email: email,
        code: code,
        newPassword: newPassword,
      );
      return await authRepository.resetPassword(request);
    } finally {
      GlobalLoading.dismiss();
    }
  }
}
